### DISCUSSION TITLE
Streaming from Dart to Rust

### DISCUSSION THREAD

**Discussion by temeddix** [fzyzcjy/flutter_rust_bridge#1214] [Ideas]
Hi, this idea might have already been considered but I wanted to ask if this feature is possible. The doc says [I can just call the rust function](https://cjycode.com/flutter_rust_bridge/feature/stream.html#what-about-streaming-from-dartflutter-to-rust).

I know that simply calling rust function and putting the value in `std::sync::mpsc::channel` at the Rust side is possible. However this cannot work in WASM because Rust's channel does not support WASM.

It will be very nice to stream directly from Dart to Rust for the sake of WASM environment. Would this be possible? Or would it be hard to implement?

**Comment by fzyzcjy** [comment]
I guess streaming from dart to rust is more like a grammar sugar, instead of something deep. Streaming from rust to dart is indeed simple: Rust just throw (possibly multiple) data to dart side - nothing more.

> Would this be possible? Or would it be hard to implement?

It is mainly 2 parts:

1. Imagine we have pure Rust (no Dart at all), and the stream "source" comes from another Rust function instead of a Dart function. How will you implement a stream?
2. Then just replace the stream "source" with a frb call from Dart. End ;)

**Comment by temeddix** [comment]
Thanks for the fast reply :)

> * Imagine we have pure Rust (no Dart at all), and the stream "source" comes from another Rust function instead of a Dart function. How will you implement a stream?
> * Then just replace the stream "source" with a frb call from Dart. End ;)

Yeah, in native apps this is the easiest way and I can use `std::sync::mpsc::channnel`. There's not even the need to implement a stream. The hard point is that in WASM, `std::sync::mpsc::channel` is not available.

Let's say that we are running the web app like this:
- Main browser thread - Dart
- Web worker 1 - Rust(Workerpool of FRB): Where the 'receiver.recv()' happens in a loop
- Web worker 2 - Rust(Workerpool of FRB)
- Web worker 3 - Rust(Workerpool of FRB)
- Web worker 4 - Rust(Workerpool of FRB)

If we call something like `putTheValueInsideRustStream()` from Dart, it will be executed inside one of the web workers 2/3/4(If not SyncReturn) because web worker 1 is occuped with an infinite loop. But in WASM, cross-communication between web workers is not possible with pure Rust. This is mainly because shared memory and atomics does not exist in WASM. Therefore we need some 'JavaScript glue' or 'ports', which is what `flutter_rust_bridge` does.

**Reply by fzyzcjy** [reply]
Hmm I see why it is complex. cc @Desdaemon who did the great Web part!

**Comment by temeddix** [comment]
But it seems that when web worker 1 is occupied with a Rust loop, JS glue won't work either... so it's complicated 🥲

**Comment by Desdaemon** [comment]
A bit of boilerplate is needed, but I did manage to verify it works with WASM. It needs the `+atomics` feature flag in particular.

```rust
pub struct Sender {
    pub inner: RustOpaque<std::sync::mpsc::Sender<i32>>,
}

impl Sender {
    pub fn send(&self, value: i32) {
        self.inner.send(value).expect("Failed to send value");
    }
}

pub struct Receiver {
    pub inner: RustOpaque<std::sync::mpsc::Receiver<i32>>,
}

impl Receiver {
    pub fn recv(&self) -> anyhow::Result<i32> {
        self.inner
            .recv()
            .map_err(|err| anyhow::anyhow!("Failed to receive value: {err}"))
    }
}

// Only for demonstration purposes
unsafe impl Send for Sender {}
unsafe impl Send for Receiver {}

pub struct MyChannel {
    pub sender: Sender,
    pub receiver: Receiver,
}

impl MyChannel {
    pub fn new() -> MyChannel {
        let (tx, rx) = std::sync::mpsc::channel();
        MyChannel {
            sender: Sender {
                inner: RustOpaque::new(tx),
            },
            receiver: Receiver {
                inner: RustOpaque::new(rx),
            },
        }
    }
}

// An unclosed sender will hang up the receiver
pub fn close_sender(sender: Sender) {}
```

To close the sender, one needs to set the inner opaque to be moved, resulting in this syntax:
```dart
MyChannel channel;
api.closeSender(sender: channel.sender..inner.move = true);
```

**Reply by temeddix** [reply]
So, with the `atomics` feature flag enabled, if I get `RustOpaque<std::sync::mpsc::Sender>` from Dart and use it to send messages to other Rust thread, it will work... if I understood correctly

**Reply by temeddix** [reply]
This works in native, not checked if it works in WASM yet.
